### PR TITLE
Signup: Add `get-dot-blog` signup flow

### PR DESCRIPTION
**Pull request by drewblaisdell**
This PR adds a `get-dot-blog` signup flow to signup, which is where users from Delphin will be redirected if they want to map their domain to a new .blog site. To use this flow, a user must be logged in and visit it with a `domain` query parameter - otherwise, they're redirected to the main flow.

**Testing**
- Log in to WordPress.com.
- Visit http://calypso.localhost:3000/start/get-dot-blog?domain=example.com where `example.com` is any domain that is mappable on WordPress.com.
- Submit the `get-dot-blog-survey` step by answering the survey.
- Assert that on `get-dot-blog-themes`, you are only presented with blog-oriented themes.
- Submit the `get-dot-blog-themes` step by selecting a theme.
- Submit the `get-dot-blog-plans` step by selecting a plan.
- Assert that the message while signup is processing reads `Awesome! Give us one minute while we set up example.com for you.`
- Assert that you land in checkout with a domain mapping for the site you provided in the second step.

- [x] Code
- [x] Product

### CODE REVIEW

File: client/signup/steps/plans/index.jsx [hunk 1]
```diff
@@ -46,9 +47,10 @@ class PlansStep extends Component {
 			processingMessage: isEmpty( cartItem )
 				? translate( 'Free plan selected' )
 				: translate( 'Adding your plan' ),
-			stepName: stepName,
-			stepSectionName: stepSectionName,
-			cartItem
+			stepName,
+			stepSectionName,
+			cartItem,
+			...additionalStepData
```
Comment (scruffian) [inline client/signup/steps/plans/index.jsx]: I'm unsure why to need to add both `additionalStepData` and `... additionalStepData`.

**Comment by scruffian** [comment]
Code looks great 👍 

I tested and it works as expected too, but I'll let someone else test more thoroughly.

**Comment by fabianapsimoes** [comment]
In the `survey` step, we say "WordPress.com is the best place for your WordPress blog or website". We should remove that last bit that refers to websites.

If I hit Back from the `theme` step, I'm redirected to `/start`, not `/start/get-dot-blog`.

If I hit Back from the `plans` step, I'll break the flow. Can't select a theme back in the `theme` step --- Error in console is `Uncaught TypeError: Cannot read property 'domain' of undefined`.

We should probably handle the case where a domain already in use is passed to the flow (though doing that in a separate PR is fine). cc @breezyskies in terms of how to address this. Right now, we just show users a checkout error saying "The domain domain.com is already mapped to another site".

> Assert that you land in checkout with a domain mapping for the site you provided in the second step.

I do have a domain mapping for the domain I entered, but no plan.

**Comment by breezyskies** [comment]
> We should probably handle the case where a domain already in use is passed to the flow (though doing that in a separate PR is fine). cc @breezyskies in terms of how to address this. Right now, we just show users a checkout error saying "The domain domain.com is already mapped to another site".

This feels like it might be too edge case for custom messaging at this point, but I might be missing the most likely scenario here. How would a user most likely hit this state?

**Comment by drewblaisdell** [comment]
Thanks for the review @fabianapsimoes and @scruffian.

> We should remove that last bit that refers to websites.

:+1: This will be done in #8447.

> If I hit Back from the `theme` step, I'm redirected to `/start`, not `/start/get-dot-blog`.

> If I hit Back from the `plans` step, I'll break the flow. Can't select a theme back in the `theme` step --- Error in console is `Uncaught TypeError: Cannot read property 'domain' of undefined`.

This was a bug for logged in users - we were redirecting them to paths that ended in a local slug (even for `en`) then stripping that slug with a hard refresh, losing the query string. I fixed it by removing the locale slug from signup URLs when the user is logged in, as Calypso will render the language in their settings regardless of the slug in this case.

> I do have a domain mapping for the domain I entered, but no plan.

Sorry, I should have caught this. I fixed it by updating `createSiteWithDomainItems` to create a site with any cart item.

**Comment by fabianapsimoes** [comment]
> How would a user most likely hit this state?

 @breezyskies Manually changing the URL is one, but quite edge case. I suppose this might also happen if a user tries to re-set up a domain. Definitely not something we need to address _now_, just making sure it's on your radar.

Product :+1:

**Comment by yurynix** [comment]
After I complete the flow, and before I complete the checkout I receive a notification:
screen shot 2016-10-06 at 14 54 16

Which is a bit confusing, especially the `Site Title` part.

**Comment by bisko** [comment]
@yurynix This is only happening on development/testing instances and is a known issue.

File: client/lib/signup/step-actions.js [hunk 2]
```diff
@@ -52,19 +61,12 @@ function addDomainItemsToCart( callback, dependencies, { domainItem, googleAppsC
...
-			}
-
-			if ( googleAppsCartItem ) {
-				newCartItems = [ ...newCartItems, googleAppsCartItem ];
-			}
-
-			if ( themeItem ) {
-				newCartItems = [ ...newCartItems, themeItem ];
-			}
+			const newCartItems = [
```
Comment (yurynix) [inline client/lib/signup/step-actions.js]: Really nice refactor IMO.

File: client/signup/steps/get-dot-blog-survey/index.jsx [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+
+	render() {
+		return (
+			<SurveyStep { ...this.props } />
+		);
+	}
+}
+
+GetDotBlogSurvey.propTypes = {
+	queryObject: PropTypes.object,
```
Comment (yurynix) [inline client/signup/steps/get-dot-blog-survey/index.jsx]: Shouldn't these be `.required` ?

**Comment by yurynix** [comment]
Code 👍  but I'm not really familiar with the setup flow mechanics, so I might've missed something.

**Comment by fabianapsimoes** [comment]
> Which is a bit confusing, especially the Site Title part.

`Site Title` is a really weird. Can we use the domain name or the .wordpress.com URL to automatically create a better site title?

**Comment by breezyskies** [comment]
> Site Title is a really weird. Can we use the domain name or the .wordpress.com URL to automatically create a better site title?

I would opt for the .blog domain here, since we've had problems in the past with the .wordpress.com URLs confusing users who purchased domains (and in this particular case, they would never have seen the subdomain until this point).

**Comment by scruffian** [comment]
Code 👍

**Comment by bisko** [comment]
@fabianapsimoes @breezyskies Please refer to this thread about *Site Title* p3Ex-1qD-p2

cc @meremagee

**Comment by breezyskies** [comment]
@bisko Discussed in Slack and decided against making that change for this flow. Will keep the default "Site Title" placeholder.

**Pull request merged**